### PR TITLE
Whisper/fix empty topic

### PR DESCRIPTION
**Pull request by b00ris**
Fixes https://github.com/ethereum/go-ethereum/issues/15810

### CODE REVIEW

File: whisper/whisperv5/api.go [hunk 1]
```diff
@@ -562,7 +562,7 @@ func (api *PublicWhisperAPI) NewMessageFilter(req Criteria) (string, error) {
 	}
 
 	if len(req.Topics) > 0 {
-		topics = make([][]byte, 1)
+		topics = make([][]byte, 0, 1)
```
Comment (dshulyak) [inline whisper/whisperv5/api.go]: it makes sense to create topics slice with the same capacity as req.Topics length, isn't it?

File: whisper/whisperv5/filter.go [hunk 2]
```diff
@@ -220,6 +220,10 @@ func matchSingleTopic(topic TopicType, bt []byte) bool {
 		bt = bt[:4]
 	}
 
+	if len(bt) < 4 {
```
Comment (dshulyak) [inline whisper/whisperv5/filter.go]: would be a good clean up if you will create a constant TopicLength and use it in this function

Comment (dshulyak) [reply whisper/whisperv5/filter.go]: it is already there, declared in the doc.go file

Comment (b00ris) [reply whisper/whisperv5/filter.go]: done

File: whisper/whisperv5/api.go (hunk 1 above)
Comment (b00ris) [reply whisper/whisperv5/api.go]: done

**Comment by b00ris** [comment]
@gballet @gluk256 Please, take a look to the PR

File: whisper/whisperv5/filter_test.go [hunk 3]
```diff
@@ -796,19 +796,52 @@ func TestVariableTopics(t *testing.T) {
...
-
-		f.Topics[4] = arr
+		env.Topic = BytesToTopic(f.Topics[i])
 		match = f.MatchEnvelope(env)
 		if !match {
 			t.Fatalf("failed MatchEnvelope symmetric with seed %d, step %d.", seed, i)
 		}
 
-		f.Topics[4][i]++
+		f.Topics[i][1]++
```
Comment (gballet) [inline whisper/whisperv5/filter_test.go]: Why not `f.Topics[i][4]++` ?

Comment (b00ris) [reply whisper/whisperv5/filter_test.go]: TopicLength is 4. So f.Topics[i][4] is out of range. 
I've added local const for this and changed value to 3(last byte).

**Comment by b00ris** [comment]
@gballet It's related to this check https://github.com/b00ris/go-ethereum/blob/6600165e27c0396ecf1aa472865cd0daccdb061e/whisper/whisperv5/filter.go#L223.
I've added this to avoid same bugs later.  
TestVariableTopics used topics with smaller length. I thougth that it's not correct. 
Correct me if I'm wrong.

**Comment by b00ris** [comment]
@gluk256 Please, take a look

File: whisper/whisperv5/filter_test.go [hunk 4]
```diff
@@ -796,19 +797,52 @@ func TestVariableTopics(t *testing.T) {
...
+func TestMatchSingleTopic_NotEquals_ReturnFalse(t *testing.T) {
+	bt := []byte("tes")
+	topic := BytesToTopic(bt)
+
+	if matchSingleTopic(topic, bt) {
+		t.FailNow()
+	}
+}
+
+func TestMatchSingleTopic_InsufficientLength_ReturnFalse(t *testing.T) {
```
Comment (PombeirP) [inline whisper/whisperv5/filter_test.go]: The test function name should be swapped with the one above.

File: whisper/whisperv6/filter_test.go [hunk 5]
```diff
@@ -820,19 +821,52 @@ func TestVariableTopics(t *testing.T) {
...
+func TestMatchSingleTopic_NotEquals_ReturnFalse(t *testing.T) {
+	bt := []byte("tes")
+	topic := BytesToTopic(bt)
+
+	if matchSingleTopic(topic, bt) {
+		t.FailNow()
+	}
+}
+
+func TestMatchSingleTopic_InsufficientLength_ReturnFalse(t *testing.T) {
```
Comment (PombeirP) [inline whisper/whisperv6/filter_test.go]: Same, the test name is incorrect.

File: whisper/whisperv5/filter.go [hunk 6]
```diff
@@ -216,8 +216,12 @@ func (f *Filter) MatchTopic(topic TopicType) bool {
 }
 
 func matchSingleTopic(topic TopicType, bt []byte) bool {
-	if len(bt) > 4 {
-		bt = bt[:4]
+	if len(bt) > TopicLength {
+		bt = bt[:TopicLength]
+	}
+
+	if len(bt) < TopicLength {
```
Comment (evgen-povt) [inline whisper/whisperv5/filter.go]: Why did you add this if? It breaks the partial topic mechanism.
